### ISSUE TITLE
SELinux policy bug

### ISSUE THREAD

**Issue by ZhaominW** [vmware/photon#1295]
### Describe the bug

The 'selinux-policy' is missing some predefined rules, make iptables and sshd unusable.

If enforce is enabled it will cause ssh login to be denied, and iptables rule will be empty.

The captured syslog is as follows.

![bug](https://user-images.githubusercontent.com/100418064/155714277-17022fde-34e1-41a0-a316-2d33bbd1c0b1.PNG)

### Reproduction steps

```bash
1.Install a Photon OS v4.0 (no rpm-ostree), install SELinux according to the official documentation, and reboot.
2.Create a new account and allow SSH, connect to it, and watch the system logs.
```

### Expected behavior

SSH should work fine.
iptables should not be empty.
The system log should not show avc: denied.

### Additional context

To make it easier to enable SELinux, you can also use the small script I prepared.
```
#!/bin/sh
# Zhaomin W. <coreos.dev@wzm.im>
ERROR="echo -en \\E[5;33m"
RESET="echo -en \\E[0;39m"

# Check Permissions
if ! ( id | grep -e '^uid=0(root)' )
then
    $ERROR && echo '[Error] Permission denied: Requires the script to be executed as root.' && $RESET
    exit 1
fi

req_reboot=false

# Install SELinux packages
if ! ( rpm -q selinux-policy )
then
    if ( rpm-ostree status )
    then
        rpm-ostree install selinux-policy
    else
        tdnf -y install selinux-policy
    fi
    req_reboot='true'
fi

# Enable SELinux security model in kernel
if ! ( cat /proc/cmdline | grep -i -e 'security=' )
then
    if ! ( cat /boot/systemd.cfg | grep -i -e '^systemd_cmdline=' | grep -i -e 'security=' )
    then
        sed -i 's/^systemd_cmdline=/systemd_cmdline=security=selinux /gi' /boot/systemd.cfg
        req_reboot='true'
    fi
fi

# Switch SELinux mode
read -t 10 -p 'Switch SELinux to enforcing mode? (y/n*): ' yn
if [ "$yn" = "y" ]
then
    if ! ( cat /etc/selinux/config | grep -i -e '^SELINUX=enforcing' )
    then
        sed -i 's/^SELINUX=permissive/SELINUX=enforcing/gi' /etc/selinux/config
        req_reboot='true'
    fi
else
    if ! ( cat /etc/selinux/config | grep -i -e '^SELINUX=permissive' )
    then
        sed -i 's/^SELINUX=enforcing/SELINUX=permissive/gi' /etc/selinux/config
        req_reboot='true'
    fi
fi

# Reboot
if [ "$req_reboot" = "true" ]
then
    read -t 10 -p 'Reboot now? (y/n*): ' yn
    if [ "$yn" = "y" ]
    then
        reboot
    fi
fi
```

**Comment by YustasSwamp** [comment]
Thanks for the report!

**Comment by YustasSwamp** [comment]
Hi @ZhaominW 
I'm able to reproduce `motdgen` issue with latest Photon OS 4.0 build and working on the fix.

But standard iptables rules set/unset through iptables systemd service works fine to me.
I see in avc error log there are read/write pipefs operations by iptables process. Can you provide the way or scripts how do you configure iptables.

Thanks!

**Comment by ZhaominW** [comment]
Hi @YustasSwamp,

I understand what you are saying, iptables does start properly, it does remove and add rules and they take effect.

The problem is that iptables needs to read the configuration from **/etc/systemd/scripts/ip4save** and **/etc/systemd/scripts/ip6save**
If that fails, then whenever system reboot or systemctl restart iptables, we will get an empty firewall.

Steps to reproduce this issue.
Just set selinux to enforcing mode and reboot.
Use **iptables -L**
You can observe that the table is empty.


![1](https://user-images.githubusercontent.com/100418064/156161380-0301ae7c-6a12-4095-99e3-f947c38caa46.PNG)
![2](https://user-images.githubusercontent.com/100418064/156161341-9cbc7751-efe3-4176-bbe8-25f9f2eef76d.PNG)



```bash
if ! ( cat /etc/selinux/config | grep -i -e '^SELINUX=enforcing' )
then
    echo "Switch to enforcing mode, reboot now."
    sed -i 's/^SELINUX=permissive/SELINUX=enforcing/gi' /etc/selinux/config
    reboot
fi


echo "Let's look at the state of iptables"
iptables -L
echo
echo
echo


echo "Add some rules，it works"
iptables --append INPUT --protocol tcp --dport 80 --jump ACCEPT
iptables -L
echo
echo
echo


echo "restart iptables"
systemctl restart iptables
iptables -L
```

Thx.

**Comment by ZhaominW** [comment]
Hi @YustasSwamp

I understand what you mean, iptables does start properly and it does remove and add rules and take effect.

The problem is that iptables needs to read the configuration from **/etc/systemd/scripts/ip4save** and **/etc/systemd/scripts/ip6save** whenever photonos starts.
If it fails, then whenever **photon os reboots**, we get an empty iptables rules.

A freshly installed photon system will predefine some firewall policies in these two files. Block all inbound, allow all outbound, allow tcp 22.

Steps to reproduce this problem.
Simply set selinux to force mode and reboot.
Use **iptables -L**
You can observe that the table is empty.

Add some iptalbes rules, save them and **reboot the system** and it's empty.
systemctl reboot iptables can not reproduce the problem, you must reboot the system.

```bash
if ! ( cat /etc/selinux/config | grep -i -e '^SELINUX=enforcing' )
then
    echo "Switch to enforcing mode, reboot now."
    sed -i 's/^SELINUX=permissive/SELINUX=enforcing/gi' /etc/selinux/config
    reboot
fi

iptables --append INPUT --protocol tcp --dport 80 --jump ACCEPT
iptables --append INPUT --protocol tcp --dport 443 --jump ACCEPT
iptables-save >/etc/systemd/scripts/ip4save
ip6tables-save >/etc/systemd/scripts/ip6save

iptables -L
```

**Comment by ZhaominW** [comment]
Hi @YustasSwamp

I understand what you mean.
The iptables does start normally and can remove and add rules.
Even `systemctl restart iptables`, everything works fine.


But the problem occurs after the OS reboot.
The script **/etc/systemd/scripts/iptables** is run when the OS starts.
This script performs some initialization of iptables, notably at the end of the script, which imports the iptables rules.
```sh
#restore ipv4 rules
iptables-restore $WAIT_OPTIONS < /etc/systemd/scripts/ip4save

#restore ipv6 rules
ip6tables-restore $WAIT_OPTIONS < /etc/systemd/scripts/ip6save

# End /etc/systemd/scripts/iptables
```

**If the above process fails, then whenever photon os restarts, the iptables rule will be empty.**


A new installation of photon os will have some firewall policies predefined in the two files above: block all inbound, allow all outbound, allow tcp 22.

If selinux enforcing is not enabled, then iptables -L will see these rules.
Once selinux enforcing is enabled, they are gone.


Steps to reproduce the problem.
Simply set selinux to enforcing mode and restart.
Use **iptables -L**
You can observe that is empty.

Add some iptalbes rules, save and **reboot the system** it is empty.
p.s. systemctl reboot iptables does not reproduce the problem, you have to reboot the system.

```sh
If ! ( cat /etc/selinux/config | grep -i -e '^SELINUX=enforcing' )
then
    echo "Switching to forced mode, rebooting now."
    sed -i 's/^SELINUX=permissive/SELINUX=enforcing/gi' /etc/selinux/config
    reboot
fi

iptables --append INPUT --protocol tcp --dport 80 --jump ACCEPT
iptables --append INPUT --protocol tcp --dport 443 --jump ACCEPT
iptables-save >/etc/systemd/scripts/ip4save
ip6tables-save >/etc/systemd/scripts/ip6save

iptables -L

echo "Once you reboot the system, these rules are gone."
```

Thx.

**Comment by ZhaominW** [comment]
I have not made any changes or configured iptables.

/etc/systemd/scripts/iptables
/etc/systemd/scripts/ip4save
/etc/systemd/scripts/ip6save

All of the above scripts come with the Photon OS by default.

**Comment by YustasSwamp** [comment]
Thanks @ZhaominW , I'm able to reproduce iptables errors by switching to generic kernel. Fix will be delivered soon!

**Comment by ZhaominW** [comment]
Hi @YustasSwamp 

I observed that you have solved the problem by commit `229b58e599a5521c876c3c7f65625abd849dda0e`.

Thank you very much for providing such a great product!

Regards,

**Comment by YustasSwamp** [comment]
Yep, forgot to mention it here. Thanks for the valuable report!
